modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceDeploymentManager.java

```diff
@@ -0,0 +1,534 @@
+ */
+public class ServiceDeploymentManager {
+    /** Busy lock. */
+    private final GridSpinBusyLock busyLock = new GridSpinBusyLock();
+
+    /** Services discovery messages listener. */
+    private final DiscoveryEventListener discoLsnr = new ServiceDiscoveryListener();
+
+    /** Services communication messages listener. */
+    private final GridMessageListener commLsnr = new ServiceCommunicationListener();
+
+    /** Services deployments tasks. */
+    private final Map<ServiceDeploymentProcessId, ServiceDeploymentTask> tasks = new ConcurrentHashMap<>();
+
+    /** Discovery events received while cluster state transition was in progress. */
+    private final List<PendingEventHolder> pendingEvts = new ArrayList<>();
+
+    /** Topology version of latest deployment task's event. */
+    private final AtomicReference<AffinityTopologyVersion> readyTopVer =
+        new AtomicReference<>(AffinityTopologyVersion.NONE);
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** Deployment worker. */
+    private final ServicesDeploymentWorker depWorker;
+
+    /** Default dump operation limit. */
+    private final long dfltDumpTimeoutLimit;
+
+    /**
+     * @param ctx Grid kernal context.
+     */
+    protected ServiceDeploymentManager(@NotNull GridKernalContext ctx) {
+        this.ctx = ctx;
+
+        log = ctx.log(getClass());
+
+        ctx.event().addDiscoveryEventListener(discoLsnr,
+            EVT_NODE_JOINED, EVT_NODE_LEFT, EVT_NODE_FAILED, EVT_DISCOVERY_CUSTOM_EVT);
+
+        ctx.io().addMessageListener(TOPIC_SERVICES, commLsnr);
+
+        depWorker = new ServicesDeploymentWorker();
+
+        long limit = getLong(IGNITE_LONG_OPERATIONS_DUMP_TIMEOUT_LIMIT, 0);
+
+        dfltDumpTimeoutLimit = limit <= 0 ? 30 * 60_000 : limit;
+    }
+
+    /**
+     * Starts processing of services deployments tasks.
+     */
+    protected void startProcessing() {
+        assert depWorker.runner() == null : "Method shouldn't be called twice during lifecycle;";
+
+        new IgniteThread(ctx.igniteInstanceName(), "services-deployment-worker", depWorker).start();
+    }
+
+    /**
+     * Stops processing of services deployments tasks.
+     *
+     * @param stopErr Cause error of deployment manager stop.
+     */
+    protected void stopProcessing(IgniteCheckedException stopErr) {
+        try {
+            busyLock.block(); // Will not release it.
+
+            ctx.event().removeDiscoveryEventListener(discoLsnr);
+
+            ctx.io().removeMessageListener(commLsnr);
+
+            U.cancel(depWorker);
+
+            U.join(depWorker, log);
+
+            depWorker.tasksQueue.clear();
+
+            pendingEvts.clear();
+
+            tasks.values().forEach(t -> t.completeError(stopErr));
+
+            tasks.clear();
+        }
+        catch (Exception e) {
+            log.error("Error occurred during stopping deployment worker.", e);
```

Comment:
I think we would want to propagate the exception upwards, probably killing the node. Swallowing an exception doesn't seem to be good.

Reply:
Fixed.